Add clustering algorithm for parentheticals

ProbablyFaiz opened the pull request:
Closes #1931.

This PR implements the clustering algorithm described in #1931, as well as associated tests and dependencies. This PR does *not* contain frontend changes that use this new clustering algorithm; those aren't too hard, but we'll hold those back until we have an idea of what this looks like post-review.

mlissner commented on cl/citations/tests.py:
```diff
@@ -1085,3 +1093,267 @@ def _print_failed_cases(
+        ]
+        for i, groups in enumerate(expected_groups):
+            with self.subTest(f"Testing {groups} are grouped correctly.", i=i):
+                flat = sum(groups, start=[])
+                output_groups = get_parenthetical_groups(flat)
```
This is a clever approach, but it took me a while to figure out what you're doing. The test cases have things grouped properly and so you ungroup them by flattening them, then you regroup them using your code. Let's get a comment or two explaining this for me the next time I read this code, so I don't have to figure it out again.

varun-iyer replied:
Will do!

ProbablyFaiz replied:
Agree, this might be a bit too clever for its own good. Maybe a bit more readable would be:

```suggestion
                flat = list(chain.from_iterable(groups))
                output_groups = get_parenthetical_groups(flat)
```

(`from itertools import chain`)

mlissner commented on cl/citations/tests.py:
```diff
@@ -1085,3 +1093,267 @@ def _print_failed_cases(
+            ),
+        ]
+        for i, groups in enumerate(expected_groups):
+            with self.subTest(f"Testing {groups} are grouped correctly.", i=i):
```
Does this make ridiculous log lines with the entire group printed, since you put `{groups}` in there?

varun-iyer replied:
No --- I defined the `__str__` and `__repr__` of `groups` to only print `DummyParenthetical.id`, which keeps the printed length of `groups` short, and it's helpful to see group info in debug. The consequence is that you'd have to go back to the tests file to see what text corresponded to a particular ID. But my guess was that if tests are failing, you'd be doing something like that anyway, and it was the most printable way I could think of to communicate the most information I could about the test.

ProbablyFaiz replied:
My fault: removed `__str__` and `__repr__` definitions without realizing they were being implicitly relied upon here

mlissner commented on cl/citations/tests.py:
```diff
@@ -1085,3 +1093,267 @@ def _print_failed_cases(
+            DummyParenthetical(id=7, text="par7", score=1),
+        ]
+        test_pairs = [
+            ((parentheticals[0:3], simgraph), parentheticals[0]),
```
A word here about why parenthetical[0] won would be helpful. My interpretation is that it's because zero is connected to 3, which connects to zero and 1. One connects to 2, 3, 7, and 2 connects to 1, we already counted 3, and 7 connects back to 1 that we already counted.

In other words:

```
┌─────┐
│     │
├───┐ │ ┌─┐
│   │ │ │ │
0 1 2 3 4 5 6 7
  │ │ │       │
  │ │ │       │
  ├─┘ │       │
  │   │       │
  ├───┘       │
  │           │
  └───────────┘
```

But that's hard to see just looking at the data. My comment is pretty not great (but the graph is so-so!). Maybe there's a way to make this more intuitive?

varun-iyer replied:
Whoa, I'm not even sure "0" is supposed to win there --- let me double check these tests.

ProbablyFaiz replied:
This behavior isn't a bug - it just looks weird because of how the test is constructed. Remember that `get_representative_parenthetical` only looks at the top 20% (rounded up to the nearest integer) of parentheticals passed in. 20% of 3 rounded up is 1, and so the algorithm only ever looks at the first parenthetical.

> But that's hard to see just looking at the data. My comment is pretty not great (but the graph is so-so!). Maybe there's a way to make this more intuitive?

This is a good thought, but this part of the algorithm is actually much simpler than this: we just rank by the number of immediate neighbors each node has. See above for the explanation as to why this test gives the answer it does.

varun-iyer replied:
Oh oops --- I definitely knew that while writing the test and forgot in between,  point in case for comments :sweat:

mlissner reviewed (changes requested):
OK! This looks really good, but before we merge it, a couple things:

1. I really appreciate that you included nice docstrings in some functions, but unfortunately they're not our house style. We usually use Sphinx style: https://www.sphinx-doc.org/en/master/usage/restructuredtext/domains.html#info-field-lists. (The big difference is we do `:param foo:` instead of `@param foo:`.)

2. Relatedly, can you please add docstrings to all methods that lack them, both in tests and elsewhere?

3. And in a similar vein, I think a file-level docstring would be appropriate here, explaining how the functions work together and what the big picture is. I know we have the issue as pretty good documentation (and this PR), but there's a fair bit of complicated stuff here. When others come around later, having a full explanation will be helpful.

4. This is some fancy code. For the graph stuff, did you consider using networkx or igraph? I think we have both as dependencies already.

The only other piece that's missing here is how this fits in with the rest of the project. When it's run, how it's run, etc. 

Do you guys want to do a separate PR to be merged serially with that in it too, or maybe include that here also? It'd flesh out the full picture either way.

ProbablyFaiz commented on pyproject.toml:
```diff
@@ -92,6 +92,10 @@ reporters-db = "^3.2.17"
 django-waffle = "^2.1.0"
 nameparser = "^1.0.6"
 eyecite = "^2.3.0"
+scipy = "^1.8.0"
+nltk = "^3.7"
+numpy = "^1.22.2"
+datasketch = "^1.5.7"
```
Datasketch's version bound for numpy is pretty broad, and was causing Poetry to try to install an older version of numpy that, at least on my ARM-based Mac, failed to install. I specified these explicitly to get around that

ProbablyFaiz commented:
> OK! This looks really good, but before we merge it, a couple things:

1-3: will do.

Re: 4, our graph operations at the moment are relatively rudimentary, so I think breaking out the big guns would probably make the code messier (plus slower). If we start doing more complex graph operations, then it may be worth it but for now I think it's okay.

> The only other piece that's missing here is how this fits in with the rest of the project. When it's run, how it's run, etc.

> Do you guys want to do a separate PR to be merged serially with that in it too, or maybe include that here also? It'd flesh out the full picture either way.

Yeah, we can push that onto this PR. We wanted to keep this one slim initially just in case we changed up the design post-review but I think we have a good enough idea of everything to add the frontend changes on. For now, we're just going to compute the clusters at page-load time, with some caching to avoid doing it over and over.

ProbablyFaiz commented:
I spent some time this morning profiling the algorithm's performance on large numbers of parentheticals. I've sped the algorithm up around 15x with the following changes:

- Using python's `deepcopy` method instead of doing expensive object initializations over and over.
- Removing stemming of text tokens. This is a nice-to-have that has a small quality benefit but is unacceptably slow at scale.
- Tokenizing with nltk's `wordpunct_tokenize` (Regex-based) instead of `word_tokenize` (slow).

The algorithm now takes approximately 1.5 seconds to run on 1000 parentheticals (previously around 15 s), with all of the avoidable bottlenecks eliminated.

ProbablyFaiz commented on cl/opinion_page/templates/view_opinion_summaries.html:
```diff
@@ -17,45 +18,78 @@
-        </h2>
+  <div class="col-sm-9">
+    <h2>
+      <a href="{{ cluster.get_absolute_url }}?{{ request.META.QUERY_STRING }}"
```
This is pre-existing (just indented a layer deeper) and comes from trusted data, so resolving this.

ProbablyFaiz commented:
Frontend changes!

Here's the new sidebar:

<img width="303" alt="Screen Shot 2022-03-04 at 6 22 30 PM" src="https://user-images.githubusercontent.com/34611522/156812272-ca0614cf-af0c-4088-be37-21ebd320d5dc.png">

Curious on your thoughts about the "and x similar summaries" text.

Here's a short video showing the new summaries page UI:
https://user-images.githubusercontent.com/34611522/156811691-63a407be-3646-469c-89f3-e98b5817c6ae.mov

It is mildly crappy (UX is not my strong suit), and I am open to alternate suggestions for how best to lay this out. But it works.

Still working on addressing some of the other code-review comments but that's coming through shortly.

ProbablyFaiz commented:
Alrighty, I think everything up to now has been addressed and this is ready for another look.

mlissner commented on cl/lib/search_utils.py:
```diff
@@ -1031,6 +1035,36 @@ def add_depth_counts(
+
+    cache_key = f"parenthetical_groups:{cluster.pk}"
+    cache = caches["db_cache"]
+    cached_groups, num_parentheticals = cache.get(cache_key, (None, None))
+    if (
+        cached_groups is not None
+        # Groups will be recomputed when new parentheticals are added
+        and num_parentheticals == cluster.parentheticals.count()
+    ):
+        return cached_groups
+
+    # If the case has more than 1,000 parentheticals, only use the top 1,000
+    parentheticals = list(cluster.parentheticals[:1000])
+    groups = get_parenthetical_groups(parentheticals)
+    a_week = 60 * 60 * 24 * 7
+    cache.set(cache_key, (groups, len(parentheticals)), a_week)
```
So I think it's right to use Django's DB cache here since these groups will take up space. I appreciate that decision. 

But I'm actually not sure there's much point in expiring the cache at all. The funny thing about django's DB cache is that items are only culled from it when you run `cache.get` on them. If you never do that, items never get culled because, as you probably know, relational databases like postgresql don't have ttl primitives built into them. (Using an in-memory DB like redis is another story.)

So put another way, once you cache something in the DB, if you have a get-check-expired-set-again pattern like above, that key is going to take up space in the DB forever. It'll take up space when you set it the first time, then it'll still be there after it expires, and someday when you learn it expired (by doing a `get`), the very next thing you do after the `get` culls it is set it again with fresh data.

In my mind, the only reasons to clear a cache are:
 - The data is old (but we check for this via parentheticals.count); or
 - To save space in our cache (but we never get this space back as explained above).

The only other reason we might want to expire it is if our scoring algo changes, but I think I'd argue that we'd want to clear the cache at that point anyway.

So with all that in mind, let's just not expire them unless y'all can think of another reason to. Or if we must, let's set really long ttls of a month or two?

ProbablyFaiz replied:
Yeah, good call. I added the count checking after the caching, and it didn't occur to me that that obviated the need for expiring the cache.

mlissner commented on cl/lib/search_utils.py:
```diff
@@ -1031,6 +1035,36 @@ def add_depth_counts(
+
+    cache_key = f"parenthetical_groups:{cluster.pk}"
+    cache = caches["db_cache"]
+    cached_groups, num_parentheticals = cache.get(cache_key, (None, None))
+    if (
+        cached_groups is not None
+        # Groups will be recomputed when new parentheticals are added
+        and num_parentheticals == cluster.parentheticals.count()
+    ):
+        return cached_groups
+
+    # If the case has more than 1,000 parentheticals, only use the top 1,000
+    parentheticals = list(cluster.parentheticals[:1000])
+    groups = get_parenthetical_groups(parentheticals)
+    a_week = 60 * 60 * 24 * 7
+    cache.set(cache_key, (groups, len(parentheticals)), a_week)
```
This needs to use len(parentheticals.count()) right? Otherwise, it will always say 1,000 for big cases? 
```
    cache.set(cache_key, (groups, len(parentheticals)), a_week)
```

ProbablyFaiz replied:
True, good catch.

mlissner commented on cl/opinion_page/views.py:
```diff
@@ -39,6 +39,7 @@
 from cl.lib.ratelimiter import ratelimit_if_not_whitelisted
 from cl.lib.search_utils import (
     get_citing_clusters_with_cache,
+    get_parenthetical_groups_with_cache,
```
Hm, I'm not sure the right place for this, but `search_utils` usually has something to do with searching. Maybe we rename `filter_parentheticals.py` to be `parenthetical_utils` or something, and put this in there?

ProbablyFaiz replied:
`search_utils` does seem to have a number of methods not related to search (citing opinions, related clusters, etc.). We can make this change, but this method is arguably much more conceptually similar to the methods around it in this file -- it computes, with caching, some related piece of information about a given opinion that were going to display -- than the parenthetical functionality.

Maybe we should move a bunch of the opinion page-related helpers in search_utils into a new utils file?

mlissner replied:
Go figure. We can get back to this some other time. Doesn't have to be here.

mlissner commented on cl/opinion_page/templates/view_opinion_summaries.html:
```diff
@@ -17,45 +18,78 @@
+                    <button class="toggle-group-summaries btn btn-link" type="button" data-toggle="collapse"
+                            aria-expanded="false" aria-controls="group-summaries"
+                            data-target="#group-summaries-{{ representative.pk }}">
+                      <span class="collapsed">See all {{ group.size|intcomma }} parentheticals like this</span>
```
```suggestion
                      <span class="collapsed">See all {{ group.size|intcomma }} cases summarizing this one</span>
```

mlissner commented on cl/opinion_page/templates/view_opinion.html:
```diff
@@ -90,12 +90,16 @@ <h3 class="inline"><span>Your Notes</span></h3>
+                    (from <a href="{{ summary.describing_opinion.cluster.get_absolute_url }}?{{ request.META.QUERY_STRING }}">
+                      {# The below line needs to be all one line or spaces gets inserted where they shouldn't be #}
+                      {{ summary.describing_opinion.cluster.caption|safe|truncatewords:10|v_wrapper }}</a>{% if group.size > 1 %}
+                        <span>and {{ group.size }} similar summar{{ group.size|pluralize:"y,ies" }}</span>{% endif %})
```
```suggestion
                        <span>and {{ group.size|add:"-1"|intcomma }} other{{ group.size|add:"-1"|pluralize }}</span>{% endif %})
```

ProbablyFaiz replied:
My only pause with something like this is it may imply that there are 12 other summaries that say *exactly that* rather than merely similar to that and I don't want to mislead. What do you think?

mlissner replied:
I think it should be OK. Our users will figure it out because the summary page makes it more clear.

mlissner reviewed (changes requested):
This looks great and I'm eager to merge, but I did find one bug (I think), and have a few comments. 

I didn't flag it line by line, but I also saw a few more methods without docstrings and at least one docstring that still doesn't use the Sphinx format. 

I think we're basically there though, assuming my comments and suggestions all make sense.

mlissner commented:
Oh, and the screenshots and video are immensely helpful, thank you.

ProbablyFaiz commented:
Alright, I think all the docstring stuff + the other comments have been addressed, sans moving the method out of `search_utils`.

ProbablyFaiz commented:
One more thing - I got rid of `nltk` as a dependency and replaced it with a couple of hand-rolled implementations of basic text operations (see all those changes in the latest commit). I also added the `pystemmer` library in its place, which stems text literally 90x faster than `nltk`, and so makes it viable for us to stem text once again and reap the quality benefits thereof.

mlissner commented on cl/lib/search_utils.py:
```diff
@@ -1031,6 +1035,38 @@ def add_depth_counts(
+    # If the case has more than 1,000 parentheticals, only use the top 1,000
+    parentheticals = list(cluster.parentheticals[:1000])
+    groups = get_parenthetical_groups(parentheticals)
+    cache.set(cache_key, (groups, current_num_parentheticals), None)
```
```suggestion
    # Cache is in the DB (which lacks TTL primitives) and gets updated when 
    # new parentheticals are created. Thus, no need to expire it.
    cache.set(cache_key, (groups, current_num_parentheticals), None)
```

mlissner commented:
Thank you. I'm excited to get this deployed. Will report more on Slack!
